src/tracker/PatternEditorTools.cpp

```diff
@@ -1888,4 +1922,111 @@ void PatternEditorTools::slotClear(mp_ubyte* dst, pp_int32 from, pp_int32 to)
 	i++;
 }
 
+bool PatternEditorTools::moveSelection(const Position& ss, const Position& se, pp_int32 moveChannels, pp_int32 moveRows, bool clear)
+{
+	if (!PatternEditorTools::hasValidSelection(pattern, ss, se))
+		return 0;
+
+	pp_int32 selectionStartChannel;
+	pp_int32 selectionStartRow;
+	pp_int32 selectionStartInner;
+	pp_int32 selectionEndChannel;
+	pp_int32 selectionEndRow;
+	pp_int32 selectionEndInner;
 
+	if (!normalizeSelection(pattern, ss, se, 
+							selectionStartChannel, selectionStartRow, selectionStartInner,
+							selectionEndChannel, selectionEndRow,selectionEndInner))
+		return 0;
+	
+	// only entire instrument column is allowed
+	if (selectionStartInner >= 1 && selectionStartInner<=2)
+		selectionStartInner = 1;
+	if (selectionEndInner >= 1 && selectionEndInner<=2)
+		selectionEndInner = 2;
+	// only entire volume column can be selected
+	if (selectionStartInner >= 3 && selectionStartInner<=4)
+		selectionStartInner = 3;
+	if (selectionEndInner >= 3 && selectionEndInner<=4)
+		selectionEndInner = 4;
+	
+	pp_int32 selectionWidth = selectionEndChannel - selectionStartChannel + 1;
+	pp_int32 selectionHeight = selectionEndRow - selectionStartRow + 1;
+	
+	mp_sint32 slotSize = pattern->effnum * 2 + 2;
+	ASSERT(slotSize == 6);
```

Comment:
A comment explaining why we expect `slotSize == 6` would be helpful here

Reply:
This was done in PatternEditorClipBoard so it seemed like I should do the same. The purpose of `effnum` isn't documented anywhere but I might assume it goes like this:

- Milky's internal pattern format is able to represent formats other than XM, for conversion purposes
- The format always has 1 byte for "note" and 1 byte for "instrument number"
- `pattern->effnum` is used to specify how much extra 'effect' data each slot contains (in 2-byte increments)
- When representing XM, this means we use an extra 2 bytes for volume, and an extra 2 bytes for effect commands
- `slotCopy` and friends are only intended to work with the XM representation, not any other format, so we assert `slotSize == 6` as a sanity check.

This is a lot of information to convey, and not the best place for it, but I've left a short comment to explain why the check is there at least.